### PR TITLE
[PLATFORM-60] Sortable lists

### PR DESCRIPTION
**Pull request by mondoreale**
I used [`react-sortable-hoc`](https://github.com/clauderic/react-sortable-hoc) to implement 2 sortable list components:
- `SortableList` – base for anything sortable, and
- `FieldList` – more complex list where items have designated handles, and can carry pretty much any content (I used `TextInput` in my example).

nov-20-2018 14-05-42

Both available in Storybook today. Pre-order starts now! 😉

### CODE REVIEW

File: app/src/shared/components/FieldList/FieldItem/index.jsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+type Props = {
+    name: string,
+}
+
+const FieldItem = ({ name }: Props) => (
+    <div className={styles.root}>
+        <div className={styles.inner}>
+            <Handle className={styles.handle} />
+            <TextInput label="" defaultValue={name} />
```
Comment (timoxley) [inline app/src/shared/components/FieldList/FieldItem/index.jsx]: How does one listen for changes on the `TextInput`? I'm thinking perhaps `TextInput` should be `{children}`?

File: app/stories/shared.stories.jsx [hunk 2]
```diff
@@ -262,3 +269,74 @@ story('Date Picker')
...
     ))
+
+class SortableListContainer extends React.Component {
+    constructor(props) {
+        super(props)
+
+        this.state = {
+            items: Array(5).fill(true).map((v, i) => `Item #${i}${i === 0 ? ' (Drag me!)' : ''}`),
+        }
+    }
```
Comment (timoxley) [inline app/stories/shared.stories.jsx]: Remember you can do away with `constructor` entirely:
```js
class SortableListContainer extends React.Component {
    state = {
        items: Array(5).fill(true).map((v, i) => `Item #${i}${i === 0 ? ' (Drag me!)' : ''}`),
    }
    …
}
```

This is the same syntax feature (class field declarations) you're already using for `onSortEnd = ({ newIndex, oldIndex }) => {`

File: app/src/shared/components/FieldList/FieldItem/index.jsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        </div>
+    </div>
+)
+
+FieldItem.styles = styles
```
Comment (timoxley) [inline app/src/shared/components/FieldList/FieldItem/index.jsx]: 👍

File: app/src/shared/components/SortableList/index.jsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+type Props = {
+    children: React.Node,
+}
+
+const SortableList = ({ children, ...props }: Props) => (
+    <div {...props}>
+        {React.Children.map(children, (child, index) => (
+            <SortableItem key={`item-${index}`} index={index}>
```
[the remark this answers is not in the document]
Comment (mondoreale) [reply app/src/shared/components/SortableList/index.jsx]: Actually… `key` comes from my previous internal iteration where I mapped an array to `SortableItem` instances (where `key` is required). It isn't for `React.Children.map`. Remove it.

File: app/src/shared/components/FieldList/FieldItem/index.jsx (hunk 1 above)
Comment (mondoreale) [reply app/src/shared/components/FieldList/FieldItem/index.jsx]: Just an example, yes.

**Comment by mondoreale** [comment]
@timoxley thanks for the feedback! 🤠 

I rebased on `development`. Fixed few issues during a rebase (see 0abfc6e6d60c678d79ec6e6efc7c02f8825616dc). The rest got addressed in 
1e9cca3...bec85fd. Cheers!

**Pull request merged**